src/meta/src/stream/scale.rs

```diff
@@ -90,6 +96,182 @@ impl RescheduleContext {
+/// being the source, and move the vnode to the dest at the end of the queue.
+///
+/// This can handle scale in, scale out, migration, and simultaneous scaling with as much affinity
+/// as possible (still needs to be tested)
+///
+/// The return value is the bitmap distribution after scaling, which covers all virtual node indexes
+pub(crate) fn rebalance_actor_vnode(
+    actors: &[StreamActor],
+    actors_to_remove: &BTreeSet<ActorId>,
+    actors_to_create: &BTreeSet<ActorId>,
+) -> HashMap<ActorId, Bitmap> {
+    assert!(actors.len() > actors_to_remove.len());
+
+    let target_actor_count = actors.len() - actors_to_remove.len() + actors_to_create.len();
+    assert!(target_actor_count > 0);
+
+    // represents the balance of each actor, used to sort later
+    struct Balance {
+        actor_id: ActorId,
+        balance: i32,
+        builder: BitmapBuilder,
+    }
+
+    let (expected, mut remain) = VIRTUAL_NODE_COUNT.div_rem(&target_actor_count);
+
+    tracing::debug!(
+        "expected {}, remain {}, prev actors {}, target actors {}",
+        expected,
+        remain,
+        actors.len(),
+        target_actor_count,
+    );
+
+    let (mut removed, mut rest): (Vec<_>, Vec<_>) = actors
+        .iter()
+        .filter_map(|actor| {
+            actor
+                .vnode_bitmap
+                .as_ref()
+                .map(|buffer| (actor.actor_id as ActorId, Bitmap::from(buffer)))
+        })
+        .partition(|(actor_id, _)| actors_to_remove.contains(actor_id));
+
+    let order_by_bitmap_desc =
+        |(_, bitmap_a): &(ActorId, Bitmap), (_, bitmap_b): &(ActorId, Bitmap)| -> Ordering {
+            bitmap_a
+                .num_high_bits()
+                .cmp(&bitmap_b.num_high_bits())
+                .reverse()
+        };
+
+    let builder_from_bitmap = |bitmap: &Bitmap| -> BitmapBuilder {
+        let mut builder = BitmapBuilder::default();
+        builder.append_bitmap(bitmap);
+        builder
+    };
+
+    removed.sort_by(order_by_bitmap_desc);
+    rest.sort_by(order_by_bitmap_desc);
+
+    let removed_balances = removed.into_iter().map(|(actor_id, bitmap)| Balance {
+        actor_id,
+        balance: bitmap.num_high_bits() as i32,
+        builder: builder_from_bitmap(&bitmap),
+    });
+
+    let mut rest_balances = rest
+        .into_iter()
+        .map(|(actor_id, bitmap)| Balance {
+            actor_id,
+            balance: bitmap.num_high_bits() as i32 - expected as i32,
+            builder: builder_from_bitmap(&bitmap),
+        })
+        .collect_vec();
+
+    let mut created_balances = actors_to_create
+        .iter()
+        .map(|actor_id| Balance {
+            actor_id: *actor_id,
+            balance: -(expected as i32),
+            builder: BitmapBuilder::zeroed(VIRTUAL_NODE_COUNT),
+        })
+        .collect_vec();
+
+    for balance in created_balances
+        .iter_mut()
+        .rev()
```

Comment:
The front of this queue is the removed actor, queued in descending order of balance, and the tail is the created actor s, which needs to be assigned the extra 1 from the tail to these created actors, so that in the case of migration,there will not be 1 remaining vnode of the deleted node.

For example, when we have 5 actors assigned 256 vnodes, there will definitely be 1 actor with 52 vnodes (the rest are 51), When we do a -1+1 migration, since the average vnode per actor is still 51 after the migration, we need to assign that extra 1 to the migrated new actor, otherwise there will be a situation where 51 are assigned to the new actor and the remaining 1 to the remain old actor, which defeats the purpose of migration